lib/vm-agent.js

```diff
@@ -9,450 +9,696 @@
+    }
 
-    function onRetry(err) {
-        if (err) {
-            log.error(err, 'Failed retry-backoff for initial sendFullSample');
+    self.queue = vasync.queue(function _updateVmapiVm(vm_uuid, callback) {
+        // closure so self is correct here
+        if (!self.ready) {
+            callback();
             return;
         }
+        self.updateVmapiVm(vm_uuid, callback);
+    }, 1);
 
-        log.info('Initial VMs state was successfully sent. Good to go');
-        self.eventWatcher.updateState(self.sample);
-        self.eventWatcher.lastCfgEvent = self.lastFullSample;
-        self.eventWatcher.start();
-    }
+    // set values to defaults
+    self.updateDelay = INITIAL_UPDATE_DELAY;
+    self.dirtyVms = [];
+    self.retryDelays = {};
+    self.ready = false;
+    self.lastSeenVms = {};
 };
 
-
-/*
- * Initializes the EventWatcher event listeners
- */
-VmAgent.prototype.initializeEventWatcher = function () {
+VmAgent.prototype.updateVmapiVm = function updateVmapiVm(vm_uuid, callback) {
     var self = this;
-    var log = this.log;
-    var eventWatcher = this.eventWatcher;
-
-    eventWatcher.on('state', function (uuid, state) {
-        log.debug('state event for %s state: %s', uuid, state);
-        self.pushSample({
-            uuid: uuid,
-            cachedVm: self.sample[uuid]
-        });
-    });
-
-    eventWatcher.on('zone_state', function (uuid, zone_state) {
-        log.debug('zone_state event for %s newstate: %s', uuid, zone_state);
-        self.pushSample({
-            uuid: uuid,
-            cachedVm: self.sample[uuid]
-        });
-    });
 
-    eventWatcher.on('zone_xml', function (uuid) {
-        log.debug('fs.watch event on /etc/zones for %s', uuid);
-        self.pushSample({
-            uuid: uuid,
-            cachedVm: self.sample[uuid]
-        });
-    });
-
-    eventWatcher.on('destroyed', function (uuid) {
-        var vm = self.sample[uuid];
-        if (!vm) {
-            log.warn('VM %s appears to have gone away but ' +
-                'self.sample doesn\'t have it', uuid);
-            return;
-        }
-
-        log.info('Found a destroyed VM %s', uuid);
-        vm.state = 'destroyed';
-        self.pushSample({ uuid: uuid, cachedVm: vm });
-    });
+    assert.ok(self.ready, 'no updates until init complete');
+
+    /*
+     * The VM has either been created, modified or deleted.
+     *
+     * We do a vmadm load here and if the VM does not exist, grab the last
+     * object we have for the VM from self.lastSeenVms and do a PUT with
+     * that object but 'state' and 'zone_state' set to 'destroyed. If the VM
+     * does exist, we just PUT the object as-is and update self.lastSeenVms.
+     *
+     */
+    vasync.pipeline({arg: {}, funcs: [
+        function _loadVm(stash, cb) {
+            var opts = {log: self.log, uuid: vm_uuid};
+            var startLoad = (new Date()).getTime();
+
+            // NOTE: vmadm.load handles filtering out the do_not_inventory VMs
```

Comment:
In case the VM has `do_not_inventory` set, does that mean `vmadm.load`'s callback is called with an error object with an error code !== `'VmNotFound'`?